includes/class-amp-theme-support.php

```diff
@@ -1140,9 +1185,39 @@ public static function prepare_response( $response, $args = array() ) {
 
 		AMP_Response_Headers::send_server_timing( 'amp_dom_serialize', -$dom_serialize_start, 'AMP DOM Serialize' );
 
+		// Cache response if enabled.
+		if ( true === $args['enable_response_caching'] ) {
+			wp_cache_set( self::$response_cache_key, $response, self::RESPONSE_CACHE_GROUP, MONTH_IN_SECONDS );
+		}
+
 		return $response;
 	}
 
+	/**
+	 * Set response cache hash, the data values dictates whether a new hash key should be generated or not.
+	 *
+	 * @since 1.0
+	 *
+	 * @param array $data Data used to build hash key. A new hash key will be generated upon data value changes
+	 *                    which will eventually trigger a new cached reponse.
+	 */
+	protected static function set_response_cache_key( $data ) {
```

Comment:
I'm confused as to why this is needed. What specifically are the closures that are causing the problem? I don't think we should be calling them to obtain their value, as what if the closures require arguments? What are the closures going to be doing?

It would be better to just omit any closures. In fact, if you use `json_encode()` instead of `serialize()` then any non-serializable values (like closures) will automatically be turned into `null`. I think that would be better approach then `set_response_cache_key` here.

Reply:
If an argument callback is used in a way that would modify the response during post processing, the cache would not be regenerated even if the returned value of the callback changes which could cause issues. I don't think any argument callback are used that way at this so should be fine `wp_json_encode()`. I definitely resonate with your comment regarding the closures arguments issue.
Update the code [in this commit](https://github.com/Automattic/amp-wp/pull/1156/commits/9bcc4acfb28d87ba04b5f23c3793cce83b5deaa3).